OpenRA.Game/Renderer.cs

```diff
@@ -123,26 +128,49 @@ public void InitializeDepthBuffer(MapGrid mapGrid)
 		public void BeginFrame(int2 scroll, float zoom)
 		{
 			Context.Clear();
+
+			// Cache property results to avoid the unnecessary overhead of repeat queries
+			var resolution = Resolution;
+			var scale = WindowScale;
+
+			var backingResolution = new Size((int)(scale * resolution.Width), (int)(scale * resolution.Height));
+			var backingBufferSize = backingResolution.NextPowerOf2();
+			if (screenSprite == null || backingBufferSize != screenSprite.Sheet.Size)
+			{
+				if (screenBuffer != null)
+					screenBuffer.Dispose();
+
+				// Render into a frame buffer to support render != window sizes
+				// and reading data back for screenshots
+				screenBuffer = Context.CreateFrameBuffer(backingBufferSize);
+				var screenSheet = new Sheet(SheetType.BGRA, screenBuffer.Texture);
+
+				// Flip sprite in Y to match OpenGL's bottom-left origin
+				var screenBounds = Rectangle.FromLTRB(0, backingResolution.Height, backingResolution.Width, 0);
+				screenSprite = new Sprite(screenSheet, screenBounds, TextureChannel.RGBA);
+			}
+
+			screenBuffer.Bind();
 			SetViewportParams(scroll, zoom);
 		}
 
 		public void SetViewportParams(int2 scroll, float zoom)
 		{
+			var bufferSize = Resolution.NextPowerOf2();
```

Comment:
This is bogus.

Fix code and comment to explain: the `bufferSize` represents the `backingBufferSize` from `BeginFrame` scaled into window units - i.e. `backingBufferSize / scale` - which is only equal to `Resolution.NextPowerOf2()` when `scale` is a power of 2.

Replace with something like
```csharp
var scale = WindowScale;
var resolution = Resolution;
var bufferSize = new Size(
	(int)(Exts.NextPowerOf2((int)(scale * resolution.Width)) / scale),
	(int)(Exts.NextPowerOf2((int)(scale * resolution.Height)) / scale));
```

and try and clean up the code.